### ISSUE TITLE
hsmd: migrate to BIP32 derivation for all keys

### ISSUE THREAD

**Issue by darosior** [ElementsProject/lightning#3687]
*In the middle of writing this I started really doubting about the feasibility/tradeoffs of it.*

*Changing the derivation would require users to close their channels and overall introduce a big mess (I think) between channel using old derivation and channels using the new one. I'm not sure the benefits outweigh the complications..*

*I'm leaving the below text for context, but do you think it's worth fixing this `FIXME` at all or is it too late ? Do you have a better automagic way of updating existing channels ?*

-----

I'd like to gather some feedback on how this can get accomplished before getting started on the implementation.

### Rationale

As noted in the following `FIXME`, LND has a [nicer BIP32-based and versioned derivation](https://github.com/lightningnetwork/lnd/blob/master/keychain/derivation.go#L48). Migrating to it would allow to share the same and compatible backup methods (for example, even if controversed, I personally find BIP39 mnemonics really good for user experience).
https://github.com/ElementsProject/lightning/blob/10f47b41fa3192638442ef04d816380950cc32c9/hsmd/hsmd.c#L375-L376

On the same subject, I think it would be a good opportunity to expand the `hsmtools` helper, for example with @jb55's [clightning-dumpkeys](https://github.com/jb55/clightning-dumpkeys) along with mnemonic-related stuff. It could also be [installed (and documented..)](https://github.com/ElementsProject/lightning/issues/3492#issuecomment-583931721).

Finally, I don't know what derivation method do ACINQ and R-L use, but it would be great for user experience to all use the same method (and LND's one seems a good candidate).

## Our current derivation

We currently use unhardened derivation only for onchain stuff.

We derive keys from `m/0/0` with the root seed being :
```hkdf_sha256(hsm_secret, salt=NULL, info="bip32 seed")```

We derive our node id with : 
```node_privkey = hkdf_sha256(hsm_secret, salt=0, info="nodeid")``` 

We derive our basepoints with:
```
channel_seed = hkdf_sha256(hkdf_sha256(hsm_secret, salt=NULL, info="peer seed"), salt=peer_id | dbid, info="per-peer seed")
[funding, revocation, payment, htlc, delayed, shaseed] = expanded_hkdf_sha256(channel_seed, salt=NULL, info="c-lightning")
```
With `expanded_hkdf_sha256` which outputs a `6 * 32` bytes-long key with each 32 bytes part used (in this order) for each basepoint.

### A versioned `hsm_secret`

A version byte is preppended to the `hsm_secret` in order to "smoothly" change our derivation method should we have to.

Currently, the `hsm_secret` can be either 32 (clear) or 73 (encrypted) bytes long. We would prepand the version byte to the unencrypted version, and we would have either 33 (clear) or 74 (encrypted) bytes long `hsm_secret`s (according to [libsodium documentation](https://doc.libsodium.org/secret-key_cryptography/secretstream) `+ 1` to the message length results in `+ 1` to the cypher length for our construction).

### A new (LND's) derivation method

LND uses : 
- `m/1017'/0'/0'/0/index` for funding basepoint,
- `m/1017'/0'/1'/0/index` for revocation basepoint,
- `m/1017'/0'/2'/0/index` for htlc basepoint,
- `m/1017'/0'/3'/0/index` for payment basepoint,
- `m/1017'/0'/4'/0/index` for delayed payment basepoint,
- `m/1017'/0'/5'/0/index` for ??shased??,
- `m/1017'/0'/6'/0/index` for the node id,

-- discontinued, see first line of this issue --

**Comment by niftynei** [comment]
> Changing the derivation would require users to close their channels and overall introduce a big mess (I think) between channel using old derivation and channels using the new one. I'm not sure the benefits outweigh the complications..

Rather than having users switch over immediately, is there a way to progressively move over, such that any newly opened channel would use this BIP32 derivation scheme and any newly minted node would automatically default to it?

**Comment by darosior** [comment]
> Rather than having users switch over immediately, is there a way to progressively move over, such that any newly opened channel would use this BIP32 derivation scheme and any newly minted node would automatically default to it?

Of course, we'd not ask user to close their channels : I meant "the channels won't be updated until they close it and open a new one".

**Comment by rustyrussell** [comment]
AI cleanup of old issues.  If wrong, please open a new issue.


Triage decision summary:
Issue proposes migrating to BIP32 derivation for key generation to improve compatibility with LND and enable BIP39 mnemonics. However, the issue author themselves expressed serious doubts about feasibility due to the breaking changes required for existing channels, and the issue appears abandoned after the initial discussion without implementation or resolution.
